logicrunner/result_matcher.go

```diff
@@ -0,0 +1,103 @@
+package logicrunner
+
+import (
+	"sync"
+
+	"context"
+
+	"github.com/insolar/insolar/insolar"
+	"github.com/insolar/insolar/insolar/message"
+	"github.com/insolar/insolar/insolar/reply"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/messagebus"
+	"github.com/pkg/errors"
+)
+
+type resultsMatcher struct {
+	lr                *LogicRunner
+	lock              *sync.RWMutex
+	executionNodes    map[insolar.Reference]insolar.Reference
+	unwantedResponses map[insolar.Reference]message.ReturnResults
+}
+
+func newResultsMatcher(lr *LogicRunner) resultsMatcher {
+	return resultsMatcher{
+		lr:                lr,
+		lock:              &sync.RWMutex{},
+		executionNodes:    make(map[insolar.Reference]insolar.Reference),
+		unwantedResponses: make(map[insolar.Reference]message.ReturnResults),
+	}
+}
+
+func (rm *resultsMatcher) send(ctx context.Context, msg insolar.Message, receiver *insolar.Reference) {
+	sender := messagebus.BuildSender(
+		rm.lr.MessageBus.Send,
+		messagebus.RetryIncorrectPulse(rm.lr.PulseAccessor),
+		messagebus.RetryFlowCancelled(rm.lr.PulseAccessor),
+	)
+	_, err := sender(ctx, msg, &insolar.MessageSendOptions{
+		Receiver: receiver,
+	})
+	if err != nil {
+		inslogger.FromContext(ctx).Warn(errors.Wrap(err, "[ resultsMatcher::send ] Couldn't resend response"))
+	}
+}
+
+func (rm *resultsMatcher) AddStillExecution(ctx context.Context, msg *message.StillExecuting) {
+	rm.lock.Lock()
+	defer rm.lock.Unlock()
+
+	for _, reqRef := range msg.RequestRefs {
+		if response, ok := rm.unwantedResponses[reqRef]; ok {
+			inslogger.FromContext(ctx).Debug("[ resultsMatcher::AddStillExecution ] resend unwanted response ", reqRef)
+			go rm.send(ctx, &response, &msg.Executor)
+			delete(rm.executionNodes, reqRef)
```

Comment:
remove delete and continue, move rm.executionNodes[reqRef] = msg.Executor to else